src/rules.jl

```diff
@@ -96,3 +96,110 @@ true
 See also: [`frule`](@ref), [`@scalar_rule`](@ref)
 """
 rrule(::Any, ::Vararg{Any}; kwargs...) = nothing
+
+
+#######################################################################
+# Infastructure to support generating overloads from rules.
```

Comment:
Is this worth putting in its own file?

Reply:
idk, I kind of feel like it belongs here?
Its is still about the rules.